firewood/src/merkle/stream.rs

```diff
@@ -396,6 +396,140 @@ impl<'a, S: ShaleStore<Node> + Send + Sync, T> Stream for MerkleKeyValueStream<'
     }
 }
 
+enum PathIteratorState<'a> {
+    Iterating {
+        /// The key, as nibbles, of the node at [address], without the
+        /// node's partial path (if any) at the end.
+        /// Invariant: If this node has a parent, the parent's key is a
+        /// prefix of the key we're seeking.
+        /// Note the node at [address] may not have a key which is a
+        /// prefix of the key we're seeking.
+        matched_key: Vec<u8>,
+        unmatched_key: NibblesIterator<'a, 0>,
+        address: DiskAddress,
+    },
+    Exhausted,
+}
+
+/// Iterates over all nodes on the path to a given key starting from the root.
+/// If the key isn't in the trie, the final returned node is either the node with the
+/// longest prefix of the key or the node that is where the key would be inserted if
+/// it were in the trie. Note this means the final returned node may not be a prefix
+/// of the key we're seeking.
+pub struct PathIterator<'a, 'b, S, T> {
+    state: PathIteratorState<'b>,
+    merkle: &'a Merkle<S, T>,
+}
+
+impl<'a, 'b, S: ShaleStore<Node> + Send + Sync, T> PathIterator<'a, 'b, S, T> {
+    pub(super) fn new(
+        key: &'b [u8],
+        merkle: &'a Merkle<S, T>,
+        merkle_root: DiskAddress,
+    ) -> Result<Self, MerkleError> {
+        let sentinel_node = merkle.get_node(merkle_root)?;
```

Comment:
Instead of passing in the root, you could clean up the API by passing in the sentinel node. Then you can return `Self` instead of a `Result`. 

Doing this causes you to write a little more repeated code, but it's a lot more explicit. I would rather have the extra line that makes it explicit why there's a failure *before* calling the constructor rather than after. This pattern is especially useful in places where you can retry (which I don't think is applicable here). 

This isn't actually a Rust-specific pattern, but I've found Rust code bases (libraries or apps) a lot easier to maintain if you decouple fallible and infallible code as much as possible.